Blocks: Fix excess whitespace in block style class name (new package: TokenList)

aduth opened the pull request:
This pull request seeks to resolve an issue where assigning a block style can introduce unnecessary whitespace to its `className`. In doing so, it implements a new package, `@wordpress/token-list`, intended to imitate the interface of [`DOMTokenList`](https://developer.mozilla.org/en-US/docs/Web/API/DOMTokenList) while being constructable and not requiring a DOM. `DOMTokenList` is the interface used by [`Element#classList`](https://developer.mozilla.org/en-US/docs/Web/API/Element/classList), which is an easier mechanism for working with (adding, removing, replacing) class names.

**Testing instructions:**

Ensure unit tests pass:

```
npm test
```

Verify that changing the "Quote" block style to "Large" has no leading or trailing whitespace in its assigned class name.

aduth committed:
Packages: Add package token-list

aduth committed:
Editor: Fix block styles class name excess whitespace

aduth committed:
Packages: Add token list as client asset built package

aduth commented on packages/editor/src/components/block-styles/index.js:
```diff
@@ -55,22 +52,16 @@ export function getActiveStyle( styles, className ) {
+
+	const list = new TokenList( className );
+	if ( activeStyle ) {
+		list.replace( 'is-style-' + activeStyle.name, newStyleName );
```
It should be noted that the behavior of `DOMTokenList#replace` is such that, if the original token is not found, then the new token is not added. Meaning we _assume_ here that `activeStyle` is present, otherwise the new style name will not be included.

aduth replied:
In fact, this behavior is to blame for the regression observed through the failing end-to-end test, resolved (with new unit test case) in 6fcf8a5.

aduth committed:
Editor: Replace block style by removal

TokenList#replace does not add a style if the original does not exist, which is occasionally true for blocks which have an assumed default active style without the explicitly assigned class name

youknowriad commented on packages/token-list/src/index.js:
```diff
@@ -0,0 +1,172 @@
+/**
+ * External dependencies
+ */
+import { uniq, compact, without } from 'lodash-es';
```
Do you think we should just use `lodash` for now and update to `lodash-es` consistently across packages in a separate PR?

aduth replied:
Yes, that seems reasonable.

aduth committed:
Testing: Update (fix) style variation snapshot

aduth committed:
Packages: Use lodash instead of lodash-es

Save for future enhancement

aduth commented:
Reverted changes to introduce `lodash-es`.

Also updated the snapshot which was (legitimately) failing. Basically the entire intent of the pull request encompassed in 53320eb 😄

youknowriad commented on lib/client-assets.php:
```diff
@@ -134,6 +134,13 @@ function gutenberg_register_scripts_and_styles() {
+	wp_register_script(
+		'wp-token-list',
+		gutenberg_url( 'build/token-list/index.js' ),
+		array(),
```
Should we add `lodash` here

aduth replied:
Yes. 😅

aduth replied:
Observation: Too much human time spent toward considering adding / removing dependencies from various places (`client-assets.php`, `package.json`). Could really do for some automation / build testing.

gziolo replied:
Yes, I raised it once during Core JS meeting. We agreed we should do it, but there is no action in the area. Ideally, we could offer Webpack plugin which would generate JSON file to consume by WordPress. To make it work we would have to ensure that all deps are moved to WordPress packages though.

aduth replied:
Do you have a link handy to where it was discussed previously? Is it worth adding again to the agenda?

youknowriad reviewed (approved):
LGTM 👍

aduth committed:
Framework: Add missing dependency for TokenList
